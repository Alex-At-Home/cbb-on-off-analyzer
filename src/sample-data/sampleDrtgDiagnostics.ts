export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
