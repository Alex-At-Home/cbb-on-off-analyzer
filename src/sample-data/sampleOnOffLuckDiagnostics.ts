
/** Manually built by copying the snapshot from src/utils/stats/__tests__/LuckUtils.test.ts */
export const sampleOffOnOffLuckDiagnostics =
{
   "avgEff": 100,
   "samplePoss": 1602,
   "sample3P": 0.31958762886597936,
   "sample3PA": 582,
   "base3PA": 765,
   "player3PInfo": {
      "Cowan, Anthony": {
         "shot_info_ast_3pm": 35,
         "shot_info_early_3pa": 23,
         "shot_info_unast_3pm": 13,
         "shot_info_unknown_3pM": 103,
         "shot_info_total": 174,
         "base3P": 0.3218390804597701,
         "unassisted3P": 0.2780890804597701,
         "assisted3P": 0.3380890804597701,
         "expected3P": 0.3218390804597701
      },
      "Wiggins, Aaron": {
         "shot_info_ast_3pm": 49,
         "shot_info_early_3pa": 47,
         "shot_info_unast_3pm": 0,
         "shot_info_unknown_3pM": 71,
         "shot_info_total": 167,
         "base3P": 0.31736526946107785,
         "unassisted3P": 0.25736526946107785,
         "assisted3P": 0.31736526946107785,
         "expected3P": 0.31736526946107785
      }
   },
   "sampleBase3P": 0.3196480938416422,
   "regress3P": 0.31962196866284803,
   "sampleOff3PRate": 0.4279411764705882,
   "sampleOffFGA": 1360,
   "sampleOffOrb": 0.30924170616113744,
   "sampleOffEfg": 0.49779411764705883,
   "sampleOffPpp": 108.3645443196005,
   "sampleDefSos": 96.33476903870161,
   "delta3P": 0.0000343397968686765,
   "deltaOffEfg": 0.00002204311960761366,
   "deltaMissesPct": -0.000014695413071742441,
   "deltaOffPppNoOrb": 0.0037426520182714832,
   "deltaOffOrbFactor": -0.00000454441395925623,
   "deltaPtsOffMisses": -0.0004924703560545094,
   "deltaOffPpp": 0.003250181662216974,
   "deltaOffAdjEff": 0.0033738407167522694
};


/** Manually built by copying the snapshot from src/utils/stats/__tests__/LuckUtils.test.ts */
export const sampleDefOnOffLuckDiagnostics =
{
   "avgEff": 100,
   "luckPct": 0.66,
   "baseDef3P": 0.32521489971346706,
   "baseDef3PSos": 0.3298939828080229,
   "basePoss": 2081,
   "base3PSosAdj": -0.0015908882521489841,
   "sampleDef3P": 0.3614457831325301,
   "sampleDef3PSos": 0.3225421686746988,
   "samplePoss": 489,
   "sample3PSosAdj": 0.013227228915662642,
   "sampleDefEfg": 0.4733178654292343,
   "sampleDefPpp": 98.15950920245399,
   "sampleOffSos": 105.45848670756645,
   "sampleDef3PRate": 0.3851508120649652,
   "sampleDefFGA": 431,
   "sampleDefOrb": 0.2947761194029851,
   "avg3PSosAdj": 0.001228590072777041,
   "adjDef3P": 0.32377075874747585,
   "delta3P": -0.03767502438505427,
   "deltaDefEfg": -0.021765849354706525,
   "deltaDefPppNoOrb": -3.836842974183441,
   "deltaMissesPct": 0.014510566236471017,
   "deltaDefOrbFactor": 0.004295742880401803,
   "deltaPtsOffMisses": 0.40518592191060876,
   "deltaDefPpp": -3.4316570522728322,
   "deltaDefAdjEff": -3.2540359333893396
};
