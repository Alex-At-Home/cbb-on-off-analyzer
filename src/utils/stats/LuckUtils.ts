import _ from "lodash";

/** Whether to calculate the luck adjustment over the baseline or the entire season
    Baseline will be subject to more noise but enables luck to be reduced in cases
    where you suspect the baseline is more representative (eg injuries)
*/
export type LuckAdjustmentBaseline = "baseline" | "season";

/** Holds all the info required to calculate and explain the delta when luck is regressed away */
export type OffLuckAdjustmentDiags = {
  avgEff: number,
  samplePoss: number,
  sample3P: number,
  sample3PA: number,
  base3PA: number,
  player3PInfo: Record<string, { sample3PA: number, base3P: number }>,
  sampleBase3P: number,
  regress3P: number,

  // Effects:
  sampleOff3PRate: number,
  sampleOffFGA: number,
  sampleOffOrb: number,

  sampleOffEfg: number,
  sampleOffPpp: number,
  sampleDefSos: number,

  delta3P: number,
  deltaOffEfg: number,
  deltaMissesPct: number,
  deltaOffPppNoOrb: number,

  deltaOffOrbFactor: number,
  deltaPtsOffMisses: number,
  deltaOffPpp: number,
  deltaOffAdjEff: number
};

/** Holds all the info required to calculate and explain the delta when luck is regressed away */
export type DefLuckAdjustmentDiags = {
  avgEff: number,
  /** The % of the delta between 3PD and 3P SoS  that is luck */
  luckPct: number,

  /** Opponents' 3P% against the team */
  baseDef3P: number,
  /** Opponents' 3P% over the season */
  baseDef3PSos: number,
  basePoss: number,
  /** The ability of the team to adjust the opponents' 3P% over the season */
  base3PSosAdj: number,

  // Same stats but over the sample in question: on/off/baseline
  sampleDef3P: number,
  sampleDef3PSos: number,
  samplePoss: number
  sample3PSosAdj: number,

  // (We'll need these stats later)
  sampleDefEfg: number,
  sampleDefPpp: number,
  sampleOffSos: number,
  sampleDef3PRate: number,
  sampleDefFGA: number,
  sampleDefOrb: number,

  /** The weighted sum of the two 3PD adjustments */
  avg3PSosAdj: number,
  /** The final number */
  adjDef3P: number,

  // The effect:
  delta3P: number,
  deltaDefEfg: number,
  /** The delta in raw Pts/100 not including the ORB change */
  deltaDefPppNoOrb: number,
  deltaMissesPct: number,
  deltaDefOrbFactor: number,
  deltaPtsOffMisses: number,
  deltaDefPpp: number,
  deltaDefAdjEff: number
};

/** Contains logic to help other stats modules adjust for luck */
export class LuckUtils {

  /** Calculate the offensive luck adjustment for a team ...samplePlayers==players.map(_.on/off/baseline) */
  static readonly calcOffTeamLuckAdj = (
    sampleTeam: any, samplePlayers: any[], baseTeam: any, basePlayersMap: Record<string,any>,
    avgEff: number
  ) => {
    const get = (maybeOld: any, fallback: number) => {
      // Uses the non-adjusted luck number if present
      return (_.isNil(maybeOld?.old_value) ? maybeOld?.value : maybeOld?.old_value) || fallback;
    }

    // Number of 3P shots taken in sample

    const samplePoss = get(sampleTeam?.off_poss, 0);
    const sample3P = get(sampleTeam?.off_3p, 0);
    const sample3PA = get(sampleTeam?.total_off_3p_attempts, 0);
    const base3PA = get(baseTeam?.total_off_3p_attempts, 0);

    // Loop over sample roster - lookup into base to get 3PA shots 3P%

    var varTotal3PA = 0.0;
    var varTotal3P = 0.0;
    const player3PInfo = _.chain(samplePlayers).flatMap((player: any) => {
      const samplePlayer3PA = get(player?.total_off_3p_attempts, 0);
      const samplePlayer3P = get(basePlayersMap[player?.key]?.off_3p, 0);
      varTotal3PA += samplePlayer3PA;
      varTotal3P += samplePlayer3PA*samplePlayer3P;

      return (samplePlayer3PA > 0) ? //(don't bother with any players who didn't take a 3P shot)
        [ [ player?.key, { sample3PA: samplePlayer3PA, base3P: samplePlayer3P }  ] ] : [];

    }).sortBy((pV: any[]) => -1*(pV?.[1]?.sample3PA || 0)).fromPairs().value();

    // Calculate average weight of 3P% weighted by 3PA

    const sampleBase3P = varTotal3P / (varTotal3PA || 1);

    // Regress vs actual 3P%

    const total3PA = (sample3PA + base3PA) || 1;
    const regress3P = (sampleBase3P*base3PA + sample3P*sample3PA)/total3PA;

    // Calculate effects similarly to calcDefTeamLuckAdj
    const sampleOff3PRate = get(sampleTeam?.off_3pr, 0);
    const sampleOffFGA = get(sampleTeam?.total_off_2p_attempts, 0) + get(sampleTeam?.total_off_3p_attempts, 0);
    const sampleOffOrb = get(sampleTeam?.off_orb, 0);

    const sampleOffEfg = get(sampleTeam?.off_efg, 0);
    const sampleOffPpp = get(sampleTeam?.off_ppp, 0);
    const sampleDefSos = get(sampleTeam?.def_adj_opp, 0);

    // const threePointRate = sampleDef3PA/((sampleDef3PA + sampleDef2PA) || 1)
    const delta3P = regress3P - sample3P;
    const deltaOffEfg = 1.5*delta3P*sampleOff3PRate;
    const deltaMissesPct = -1*delta3P*sampleOff3PRate;
    const deltaOffPppNoOrb = 200*deltaOffEfg*sampleOffFGA/(samplePoss || 1);
    // pts_off_misses = delta_misses*ORB*(ppp_no_orb + pts_off_misses)
    // ie pts_off_misses = delta_misses*ORB*ppp_no_orb/(1 - delta_misses*ORB)
    const deltaOffOrbFactor = deltaMissesPct*sampleOffOrb/(1 - deltaMissesPct*sampleOffOrb);
    const deltaPtsOffMisses = deltaOffOrbFactor*(sampleOffPpp + deltaOffPppNoOrb);
    const deltaOffPpp = deltaOffPppNoOrb + deltaPtsOffMisses;
    const deltaOffAdjEff = deltaOffPpp*avgEff/(sampleDefSos || 1);

    return {
      avgEff,
      samplePoss,
      sample3P,
      sample3PA,
      base3PA,
      player3PInfo,
      sampleBase3P,
      regress3P,

      sampleOff3PRate, sampleOffFGA, sampleOffOrb,
      sampleOffEfg, sampleOffPpp, sampleDefSos,

      delta3P, deltaOffEfg, deltaMissesPct, deltaOffPppNoOrb,

      deltaOffOrbFactor, deltaPtsOffMisses, deltaOffPpp, deltaOffAdjEff
    } as OffLuckAdjustmentDiags;
  };

  /** Calculate the defensive luck adjustment for a team */
  static readonly calcDefTeamLuckAdj = (sample: any, base: any, avgEff: number) => {
    const get = (maybeOld: any, fallback: number) => {
      // Uses the non-adjusted luck number if present
      return (_.isNil(maybeOld?.old_value) ? maybeOld?.value : maybeOld?.old_value) || fallback;
    }
    const luckPct = 0.66;

    //TODO: if we don't have 3PSos then just regress
    //TODO: also add a test for this case

    const baseDef3P = get(base?.def_3p, 0);
    const baseDef3PSos = _.isNil(base?.def_3p_opp?.value) ? baseDef3P : 0.01*get(base?.def_3p_opp, 0); //(normalize to %)
    const basePoss = get(base?.def_poss, 0);
    const base3PSosAdj = (1 - luckPct)*(baseDef3P - baseDef3PSos);

    const sampleDef3P = get(sample?.def_3p, 0);
    const sampleDef3PSos = _.isNil(sample?.def_3p_opp?.value) ? sampleDef3P : 0.01*get(sample?.def_3p_opp, 0); //(normalize to %)
    const samplePoss = get(sample?.def_poss, 0);
    const sample3PSosAdj = (1 - luckPct)*(sampleDef3P - sampleDef3PSos);

    const sampleDefEfg = get(sample?.def_efg, 0);
    const sampleDefPpp = get(sample?.def_ppp, 0);
    const sampleOffSos = get(sample?.off_adj_opp, 0);

    const avg3PSosAdj = (samplePoss*sample3PSosAdj + basePoss*base3PSosAdj)/((samplePoss + basePoss) || 1);
    const adjDef3P = sampleDef3PSos + avg3PSosAdj;

    const sampleDef3PRate = get(sample?.def_3pr, 0);
    const sampleDefFGA = get(sample?.total_def_2p_attempts, 0) + get(sample?.total_def_3p_attempts, 0);
    const sampleDefOrb = get(sample?.def_orb, 0);
    //
    // const threePointRate = sampleDef3PA/((sampleDef3PA + sampleDef2PA) || 1)
    const delta3P = adjDef3P - sampleDef3P;
    const deltaDefEfg = 1.5*delta3P*sampleDef3PRate;
    const deltaMissesPct = -1*delta3P*sampleDef3PRate;
    const deltaDefPppNoOrb = 200*deltaDefEfg*sampleDefFGA/(samplePoss || 1);
    // pts_off_misses = delta_misses*ORB*(ppp_no_orb + pts_off_misses)
    // ie pts_off_misses = delta_misses*ORB*ppp_no_orb/(1 - delta_misses*ORB)
    const deltaDefOrbFactor = deltaMissesPct*sampleDefOrb/(1 - deltaMissesPct*sampleDefOrb);
    const deltaPtsOffMisses = deltaDefOrbFactor*(sampleDefPpp + deltaDefPppNoOrb);
    const deltaDefPpp = deltaDefPppNoOrb + deltaPtsOffMisses;
    const deltaDefAdjEff = deltaDefPpp*avgEff/(sampleOffSos || 1);

    return {
      avgEff, luckPct,
      baseDef3P, baseDef3PSos, basePoss,
      base3PSosAdj,

      sampleDef3P, sampleDef3PSos, samplePoss,
      sample3PSosAdj,

      sampleDefEfg, sampleDefPpp, sampleOffSos,
      sampleDef3PRate, sampleDefFGA, sampleDefOrb,

      avg3PSosAdj, adjDef3P,

      delta3P, deltaDefEfg,
      deltaDefPppNoOrb, deltaMissesPct, deltaDefOrbFactor, deltaPtsOffMisses,
      deltaDefPpp, deltaDefAdjEff
    } as DefLuckAdjustmentDiags;

  };

  /** Mutates (in a reversible way) the team stats with luck adjustments */
  static readonly injectLuck = (
    mutableStats: any,
    offLuck: OffLuckAdjustmentDiags | undefined,
    defLuck: DefLuckAdjustmentDiags | undefined
  ) => {
    const reset = (mutableVal: any) => {
      return _.isNil(mutableVal?.old_value) ? mutableVal?.value : mutableVal?.old_value;
    }

    // Offense - 3P

    const off3P = reset(mutableStats?.off_3p)
    mutableStats.off_3p = offLuck ? {
      value: off3P + offLuck.delta3P,
      old_value: off3P,
      override: "Luck adjusted"
    } : {
      value: off3P
    };

    // Offense - derived 4 factors and efficiency

    const eFgOff = reset(mutableStats?.off_efg);
    mutableStats.off_efg = offLuck ? {
      value: eFgOff + offLuck.deltaOffEfg,
      old_value: eFgOff,
      override: "Adjustment derived from Off 3P%"
    } : {
      value: eFgOff
    };

    const rawOffPpp = reset(mutableStats?.off_ppp);
    mutableStats.off_ppp = offLuck ? {
      value: rawOffPpp + offLuck.deltaOffPpp,
      old_value: rawOffPpp,
      override: "Adjustment derived from Off 3P%"
    } : {
      value: rawOffPpp
    };

    const adjOffPpp = reset(mutableStats?.off_adj_ppp);
    mutableStats.off_adj_ppp = offLuck ? {
      value: adjOffPpp + offLuck.deltaOffPpp,
      old_value: adjOffPpp,
      override: "Adjustment derived from Off 3P%"
    } : {
      value: adjOffPpp
    };

    // Defense - 3P

    const def3P = reset(mutableStats?.def_3p)
    mutableStats.def_3p = defLuck ? {
      value: defLuck.adjDef3P,
      old_value: def3P,
      override: "Luck adjusted"
    } : {
      value: def3P
    };

    // Defense - derived 4 factors and efficiency

    const eFgDef = reset(mutableStats?.def_efg);
    mutableStats.def_efg = defLuck ? {
      value: eFgDef + defLuck.deltaDefEfg,
      old_value: eFgDef,
      override: "Adjustment derived from Def 3P%"
    } : {
      value: eFgDef
    };

    const rawDefPpp = reset(mutableStats?.def_ppp);
    mutableStats.def_ppp = defLuck ? {
      value: rawDefPpp + defLuck.deltaDefPpp,
      old_value: rawDefPpp,
      override: "Adjustment derived from Def 3P%"
    } : {
      value: rawDefPpp
    };

    const adjDefPpp = reset(mutableStats?.def_adj_ppp);
    mutableStats.def_adj_ppp = defLuck ? {
      value: adjDefPpp + defLuck.deltaDefPpp,
      old_value: adjDefPpp,
      override: "Adjustment derived from Def 3P%"
    } : {
      value: adjDefPpp
    };

  }

};
