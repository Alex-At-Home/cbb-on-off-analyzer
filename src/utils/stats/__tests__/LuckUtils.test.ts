

import _ from 'lodash';

import { LuckUtils } from "../LuckUtils";
import { GameFilterParams, LineupFilterParams, TeamReportFilterParams } from "../../FilterModels";
import { samplePlayerStatsResponse } from "../../../sample-data/samplePlayerStatsResponse";
import { sampleTeamStatsResponse } from "../../../sample-data/sampleTeamStatsResponse";
import { sampleOffOnOffLuckDiagnostics, sampleDefOnOffLuckDiagnostics } from "../../../sample-data/sampleOnOffLuckDiagnostics";


describe("LuckUtils", () => {

  const baseTeam = sampleTeamStatsResponse.responses[0].aggregations.global.only.buckets.team;
  const basePlayers =
    samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets;
  const basePlayersMap = _.fromPairs(
    basePlayers.map((p: any) => [ p.key, p ])
  );
  const sampleTeamOn =
    sampleTeamStatsResponse.responses[0].aggregations.tri_filter.buckets.on;
  const samplePlayersOn =
    samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.on.player.buckets;
  const sampleTeamOff =
    sampleTeamStatsResponse.responses[0].aggregations.tri_filter.buckets.off;

  const savedSampleTeamOn = _.cloneDeep(sampleTeamOn);

  // Write the data objects out in pure JS format so we can store them
  expect.addSnapshotSerializer({
    test: (val: any) => true,
    print: (val: any) => JSON.stringify(val, null, 3)
  });

  test("LuckUtils - calcOffTeamLuckAdj", () => {
    const offTeamLuckAdj = LuckUtils.calcOffTeamLuckAdj(
      sampleTeamOn, samplePlayersOn, baseTeam, basePlayersMap, 100.0
    );
    expect(offTeamLuckAdj).toMatchSnapshot("Offensive Luck Adjustments");
    // Now copy the snapshot here:
    expect(offTeamLuckAdj).toEqual(sampleOffOnOffLuckDiagnostics);
  });
  test("LuckUtils - calcOffPlayerLuckAdj", () => {
    // (just check it's a straight translation of the team version)
    const offTeamLuckAdj = LuckUtils.calcOffTeamLuckAdj(
      samplePlayersOn[0], [ samplePlayersOn[0] ], basePlayers[0], { [basePlayers[0].key]: basePlayers[0] }, 100.0
    );
    const offPlayerLuckAdj = LuckUtils.calcOffPlayerLuckAdj(
      samplePlayersOn[0], basePlayers[0], 100.0
    );
    expect(offPlayerLuckAdj).toEqual(offTeamLuckAdj);
  });
  test("LuckUtils - calcDefTeamLuckAdj", () => {
    const defTeamLuckAdj = LuckUtils.calcDefTeamLuckAdj(
      sampleTeamOff, baseTeam, 100.0
    );
    expect(defTeamLuckAdj).toMatchSnapshot("Defensive Luck Adjustments");
    // Now copy the snapshot here:
    expect(defTeamLuckAdj).toEqual(sampleDefOnOffLuckDiagnostics);
  });
  test("LuckUtils - calcDefPlayerLuckAdj", () => {
    // (just check it's a - somewhat! - straight translation of the team version)
    // (TODO this is a bit horrible, I'm just copy/pasting the logic I'm testing, but not sure
    //  how better to test)
    const samplePlayerWithExtraStats = _.assign(_.cloneDeep(samplePlayersOn[0] as any), {
      def_3p: { value: samplePlayersOn[0].oppo_total_def_3p_made.value /  samplePlayersOn[0].oppo_total_def_3p_attempts.value },
      def_3p_opp: samplePlayersOn[0].oppo_def_3p_opp,
      def_poss: samplePlayersOn[0].oppo_total_def_poss
    });
    const basePlayerWithExtraStats = _.assign(_.cloneDeep(basePlayers[0] as any), {
      def_3p: { value: basePlayers[0].oppo_total_def_3p_made.value /  basePlayers[0].oppo_total_def_3p_attempts.value },
      def_3p_opp: basePlayers[0].oppo_def_3p_opp,
      def_poss: basePlayers[0].oppo_total_def_poss
    });
    const defTeamLuckAdj = _.assign(LuckUtils.calcDefTeamLuckAdj(
      samplePlayerWithExtraStats, basePlayerWithExtraStats, 100.0
    ), {
      sampleDefOrb: 0, //(we ignore ORBs)
      sampleOffSos: 0 //(for individual players, don't transform it into efficiency so this isn't needed)
    });
    const defPlayerLuckAdj = LuckUtils.calcDefPlayerLuckAdj(
      samplePlayersOn[0], basePlayers[0], 100.0
    );
    expect(defPlayerLuckAdj).toEqual(defTeamLuckAdj);
  });
  test("LuckUtils - injectLuck", () => {
    const offTeamLuckAdj = LuckUtils.calcOffTeamLuckAdj(
      sampleTeamOn, samplePlayersOn, baseTeam, basePlayersMap, 100.0
    );
    const defTeamLuckAdj = LuckUtils.calcDefTeamLuckAdj(
      sampleTeamOff, baseTeam, 100.0
    );

    // Check object with missing fields are preserved:
    const mutableEmpty = {};
    LuckUtils.injectLuck(mutableEmpty, offTeamLuckAdj, defTeamLuckAdj);
    expect(mutableEmpty).toEqual({});
    LuckUtils.injectLuck(mutableEmpty, undefined, undefined);
    expect(mutableEmpty).toEqual({});

    // Check diffs are about what's expected

    LuckUtils.injectLuck(sampleTeamOn, offTeamLuckAdj, defTeamLuckAdj);
    LuckUtils.injectLuck(sampleTeamOff, offTeamLuckAdj, defTeamLuckAdj);

    expect(_.fromPairs(_.differenceWith(
      _.toPairs(sampleTeamOn),
      _.toPairs(savedSampleTeamOn),
      _.isEqual
    ))).toMatchSnapshot();

    // Recalculate and check that calcOffTeamLuckAdj / calcDefTeamLuckAdj

    const offTeamLuckAdj2 = LuckUtils.calcOffTeamLuckAdj(
      sampleTeamOn, samplePlayersOn, baseTeam, basePlayersMap, 100.0
    );
    const defTeamLuckAdj2 = LuckUtils.calcDefTeamLuckAdj(
      sampleTeamOff, baseTeam, 100.0
    );
    expect(offTeamLuckAdj2).toEqual(offTeamLuckAdj);
    expect(defTeamLuckAdj2).toEqual(defTeamLuckAdj);

    // Check it's idempotent:

    const savedMutatedSampleTeamOn = _.cloneDeep(sampleTeamOn);
    LuckUtils.injectLuck(sampleTeamOn, offTeamLuckAdj, defTeamLuckAdj);
    expect(sampleTeamOn).toEqual(savedMutatedSampleTeamOn);

    // Check we can reset it:
    LuckUtils.injectLuck(sampleTeamOn, undefined, undefined);
    expect(sampleTeamOn).toEqual(savedSampleTeamOn);

    //Check injectLuck when called on individual player (diff on the defensive side)

    const samplePlayerDef = { // (only difference is this one field)
      oppo_total_def_3p_attempts: {
        value: 100
      },
      oppo_total_def_3p_made: {
        value: 25
      },
    };
    const testPlayerDef = _.cloneDeep(samplePlayerDef);
    LuckUtils.injectLuck(testPlayerDef, offTeamLuckAdj, defTeamLuckAdj);
    expect(testPlayerDef).toEqual(_.assign(_.cloneDeep(samplePlayerDef), {
      oppo_def_3p: {
        value: 0.32377075874747585,
        old_value: 0.25,
        override: "Luck adjusted"
      }
    }));
    LuckUtils.injectLuck(testPlayerDef, undefined, undefined);
    expect(testPlayerDef).toEqual(_.assign(_.cloneDeep(samplePlayerDef), {
      oppo_def_3p: {
        value: 0.25
      }
    }));
  });
});
