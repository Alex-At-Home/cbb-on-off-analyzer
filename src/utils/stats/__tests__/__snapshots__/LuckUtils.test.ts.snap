// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuckUtils LuckUtils - calcDefTeamLuckAdj: Defensive Luck Adjustments 1`] = `
{
   "avgEff": 100,
   "luckPct": 0.66,
   "baseDef3P": 0.32521489971346706,
   "baseDef3PSos": 0.3298939828080229,
   "basePoss": 2081,
   "base3PSosAdj": -0.0015908882521489841,
   "sampleDef3P": 0.3614457831325301,
   "sampleDef3PSos": 0.3225421686746988,
   "samplePoss": 489,
   "sample3PSosAdj": 0.013227228915662642,
   "sampleDefEfg": 0.4733178654292343,
   "sampleDefPpp": 98.15950920245399,
   "sampleOffSos": 105.45848670756645,
   "sampleDef3PRate": 0.3851508120649652,
   "sampleDefFGA": 431,
   "sampleDefOrb": 0.2947761194029851,
   "avg3PSosAdj": 0.001228590072777041,
   "adjDef3P": 0.32377075874747585,
   "delta3P": -0.03767502438505427,
   "deltaDefEfg": -0.021765849354706525,
   "deltaDefPppNoOrb": -3.836842974183441,
   "deltaMissesPct": 0.014510566236471017,
   "deltaDefOrbFactor": 0.004295742880401803,
   "deltaPtsOffMisses": 0.40518592191060876,
   "deltaDefPpp": -3.4316570522728322,
   "deltaDefAdjEff": -3.2540359333893396
}
`;

exports[`LuckUtils LuckUtils - calcOffTeamLuckAdj: Offensive Luck Adjustments 1`] = `
{
   "avgEff": 100,
   "samplePoss": 1602,
   "sample3P": 0.31958762886597936,
   "sample3PA": 582,
   "base3PA": 765,
   "player3PInfo": {
      "Cowan, Anthony": {
         "shot_info_ast_3pm": 35,
         "shot_info_early_3pa": 23,
         "shot_info_unast_3pm": 13,
         "shot_info_unknown_3pM": 103,
         "shot_info_total": 174,
         "base3P": 0.3218390804597701,
         "unassisted3P": 0.28433908045977013,
         "assisted3P": 0.34433908045977013,
         "expected3P": 0.3235632183908046
      },
      "Wiggins, Aaron": {
         "shot_info_ast_3pm": 49,
         "shot_info_early_3pa": 47,
         "shot_info_unast_3pm": 0,
         "shot_info_unknown_3pM": 71,
         "shot_info_total": 167,
         "base3P": 0.31736526946107785,
         "unassisted3P": 0.2618935713478703,
         "assisted3P": 0.3218935713478703,
         "expected3P": 0.3186939328889391
      }
   },
   "sampleBase3P": 0.32117855364355663,
   "regress3P": 0.3204911607552493,
   "sampleOff3PRate": 0.4279411764705882,
   "sampleOffFGA": 1360,
   "sampleOffOrb": 0.30924170616113744,
   "sampleOffEfg": 0.49779411764705883,
   "sampleOffPpp": 108.3645443196005,
   "sampleDefSos": 96.33476903870161,
   "delta3P": 0.0009035318892699484,
   "deltaOffEfg": 0.0005799877495093125,
   "deltaMissesPct": -0.00038665849967287497,
   "deltaOffPppNoOrb": 0.09847482388672471,
   "deltaOffOrbFactor": -0.00011955663864158037,
   "deltaPtsOffMisses": -0.012967473985712715,
   "deltaOffPpp": 0.085507349901012,
   "deltaOffAdjEff": 0.08876063206905101
}
`;

exports[`LuckUtils LuckUtils - injectLuck 1`] = `
{
   "off_adj_ppp": {
      "value": 115.29989828118829,
      "old_value": 115.21113764911924,
      "override": "Adjusted from Off 3P%"
   },
   "def_adj_ppp": {
      "value": 86.95817296735666,
      "old_value": 90.212208900746,
      "override": "Adjusted from Def 3P%"
   },
   "off_3p": {
      "value": 0.3204911607552493,
      "old_value": 0.31958762886597936,
      "override": "Luck adjusted"
   },
   "off_ppp": {
      "value": 108.4500516695015,
      "old_value": 108.3645443196005,
      "override": "Adjusted from Off 3P%"
   },
   "off_efg": {
      "value": 0.4983741053965681,
      "old_value": 0.49779411764705883,
      "override": "Adjusted from Off 3P%"
   },
   "def_3p": {
      "value": 0.32377075874747585,
      "old_value": 0.31390977443609025,
      "override": "Luck adjusted"
   },
   "def_ppp": {
      "value": 91.98291581204879,
      "old_value": 95.41457286432161,
      "override": "Adjusted from Def 3P%"
   },
   "def_efg": {
      "value": 0.43573415064529347,
      "old_value": 0.4575,
      "override": "Adjusted from Def 3P%"
   }
}
`;
